src/external.py

```diff
@@ -0,0 +1,147 @@
+        :py:class:`hoomd.logging.Logger` are made because the logger always requests
+        the virial flags. However, this warning can be safely ignored if the
+        pressure (tensor) is not being logged or the pressure is not of interest.
+
+    """
+
+    def __init__(self, interface):
+        super().__init__()
+
+        param_dict = ParameterDict(interface=hoomd.variant.Variant)
+        param_dict["interface"] = interface
+        param_dict.update(dict(interface=interface))
+        self._param_dict.update(param_dict)
+
+        params = TypeParameter(
+            "params",
+            "particle_types",
+            TypeParameterDict(k=float, offset=float, len_keys=1),
+        )
+        self._add_typeparam(params)
+
+    def _attach_hook(self):
+        sim = self._simulation
+
+        cpp_class = getattr(_azplugins, self._make_cpp_class_name())
+
+        self._cpp_obj = cpp_class(sim.state._cpp_sys_def, self.interface)
+
+        super()._attach_hook()
+
+    def _make_cpp_class_name(self):
+        cpp_class_name = self.__class__.__name__
+        if isinstance(self._simulation.device, hoomd.device.GPU):
+            cpp_class_name += "GPU"
+        return cpp_class_name
+
+
+class PlanarHarmonicBarrier(HarmonicBarrier):
+    r"""Planar harmonic barrier normal to *z*.
+
+    Args:
+        interface (`hoomd.variant.variant_like`) : *z* position of the interface
+
+    `PlanarHarmonicBarrier` apply a purely harmonic potential in planar geometry
+    with a normal in the :math:`+z` direction:
+
+    .. math::
+
+        U(z) = \begin{cases}
+         0 & z \le H \\
+         \dfrac{\kappa}{2} (z-H)^2 & z > H
+        \end{cases}
+
+    Here, the interface is located at height *H*, which may change with time.
+    ``interface`` specifies the nominal *H*, which may then be modified
+    per-particle-type by adding an ``offset``. :math:`\kappa` is a spring
+    constant setting the strength of the interface, similar to a surface tension.
+
+    Example::
+
+        # moving interface from H = 50. to H = 25.
+        interf = hoomd.variant.Ramp(A=50.0, B=25.0, t_start=100, t_ramp=1e6)
+        evap = hoomd.azplugins.external.PlanarHarmonicBarrier(interface=interf)
+        # small particle has diameter 1.0
+        evap.params['S'] = dict(k=50.0, offset=0.0)
+        # big particle is twice as large (diameter 2.0), so coefficient is scaled
+        evap.params['B'] = dict(k=50.0*2*2, offset=0.0)
+
+    """
+
+    pass
+
+
+class SphericalHarmonicBarrier(HarmonicBarrier):
+    r"""Spherical harmonic barrier.
+
+    Args:
+        interface (`hoomd.variant.variant_like`) : Radius of spherical interface.
+
+    `SphericalHarmonicBarrier` applies a purely harmonic potential to particles
+    near a spherical interface that pushes particles inward toward the center
+    of the sphere. The potential is truncated at its minimum
+    and has the following form:
+
+    .. math::
```

Comment:
Fix the documentation from here to the end to make it mirror the planar geometry documentation. Those changes were not applied here.